packages/server/src/api/controllers/deploy/index.ts

```diff
@@ -148,6 +163,88 @@ export async function deploymentProgress(
   }
 }
 
+export async function publishStatus(ctx: UserCtx<void, PublishStatusResponse>) {
+  if (!(await features.isEnabled(FeatureFlag.WORKSPACE_APPS))) {
+    return (ctx.body = { automations: {}, workspaceApps: {} })
+  }
+  try {
+    const prodDb = context.getProdAppDB()
+    const productionExists = await prodDb.exists()
+    type State = { automations: Automation[]; workspaceApps: WorkspaceApp[] }
+    let developmentState: State = { automations: [], workspaceApps: [] }
+    let productionState: State = { automations: [], workspaceApps: [] }
+    const updateState = async (state: State) => {
+      const [automations, workspaceApps] = await Promise.all([
+        sdk.automations.fetch(),
+        sdk.workspaceApps.fetch(),
+      ])
+      state.automations = automations
+      state.workspaceApps = workspaceApps
+    }
+
+    await context.doInAppContext(context.getDevAppId(), async () =>
+      updateState(developmentState)
+    )
+
+    if (productionExists) {
+      await context.doInAppContext(context.getProdAppId(), async () =>
+        updateState(productionState)
+      )
+    }
+
+    // Create maps of production state for quick lookup
+    const prodAutomationIds = new Set(
+      productionState.automations.map(a => a._id)
+    )
+    const prodWorkspaceAppIds = new Set(
+      productionState.workspaceApps.map(w => w._id)
+    )
+
+    // Build response maps comparing development vs production
+    const automations: Record<string, PublishStatusResource> = {}
+    for (const automation of developmentState.automations) {
+      automations[automation._id!] = {
+        published: prodAutomationIds.has(automation._id!),
+        name: automation.name,
+      }
+    }
+
+    const workspaceApps: Record<string, PublishStatusResource> = {}
+    for (const workspaceApp of developmentState.workspaceApps) {
+      workspaceApps[workspaceApp._id!] = {
+        published: prodWorkspaceAppIds.has(workspaceApp._id!),
+        name: workspaceApp.name,
+      }
+    }
+
+    if (productionExists) {
+      const metadata = await getAppMetadata({ production: true })
+      if (metadata?.lastPublishedAt) {
+        const lastPublishedAt = metadata.lastPublishedAt
+        for (const automationId of Object.keys(automations)) {
+          if (lastPublishedAt[automationId]) {
+            automations[automationId].lastPublishedAt =
+              lastPublishedAt[automationId]
+          }
+        }
+        for (const workspaceAppId of Object.keys(workspaceApps)) {
+          if (lastPublishedAt[workspaceAppId]) {
+            workspaceApps[workspaceAppId].lastPublishedAt =
+              lastPublishedAt[workspaceAppId]
+          }
+        }
+      }
+    }
+
+    ctx.body = {
+      automations,
+      workspaceApps,
+    }
+  } catch (err) {
+    ctx.throw(500, "Error fetching data for deployment status")
```

Comment:
Better to let this bubble up and be handled by the error handling middleware, that way we get stack traces in DataDog with the appropriate context. Catching and rethrowing like this throws away the stack trace and creates a new one, making it almost impossible to track down the root of the problem.